### PR TITLE
update pkarr publish to also publish to relay

### PR DESCRIPTION
**Pull request by rklaehn**
also give more builder options, and use the now official pkarr record format.

This might soon move into iroh-net, but for now let's keep it here because it has so many options.

### CODE REVIEW

File: iroh-pkarr-node-discovery/src/lib.rs [hunk 1]
```diff
@@ -76,83 +93,245 @@ impl<'a> Builder<'a> {
...
+    /// Periodically publish the node address to the DHT and relay.
+    async fn publish_loop(self, keypair: SecretKey, signed_packet: SignedPacket) {
+        let this = self;
+        let z32 = pkarr::PublicKey::try_from(*keypair.public().as_bytes())
+            .expect("valid public key")
+            .to_z32();
+        loop {
+            // initial delay. If the task gets aborted before this delay is over,
+            // we have not published anything to the DHT yet.
+            tokio::time::sleep(INITIAL_PUBLISH_DELAY).await;
```
Comment (dignifiedquire) [inline iroh-pkarr-node-discovery/src/lib.rs:181]: why not just put this outside of the loop and then skip the subtraction in the second sleep?

File: iroh-pkarr-node-discovery/src/lib.rs [hunk 2]
```diff
@@ -76,83 +93,245 @@ impl<'a> Builder<'a> {
...
+    }
+
+    /// Resolve a node id from the DHT.
+    async fn resolve_dht(
+        &self,
+        pkarr_public_key: PublicKey,
+        co: &Co<anyhow::Result<DiscoveryItem>>,
+    ) {
+        if !self.0.dht {
+            return;
```
Comment (dignifiedquire) [inline iroh-pkarr-node-discovery/src/lib.rs:273]: why not yield an error?

File: iroh-pkarr-node-discovery/src/lib.rs [hunk 3]
```diff
@@ -76,83 +93,245 @@ impl<'a> Builder<'a> {
...
+        }
+    }
+
+    async fn resolve_relay(
+        &self,
+        pkarr_public_key: PublicKey,
+        co: &Co<anyhow::Result<DiscoveryItem>>,
+    ) {
+        let Some(url) = &self.0.relay_url else {
+            return;
```
[the remark this answers is not in the document]
Comment (rklaehn) [reply iroh-pkarr-node-discovery/src/lib.rs:237]: Same as above.

File: iroh-pkarr-node-discovery/src/lib.rs (hunk 2 above)
Comment (dignifiedquire) [reply iroh-pkarr-node-discovery/src/lib.rs:273]: but if both are disabled you end up with a generator that will yield nothing without error, which seems unfortunate

Comment (rklaehn) [reply iroh-pkarr-node-discovery/src/lib.rs:273]: I think it is impossible to get this. There is something about this in the builder.

Comment (dignifiedquire) [reply iroh-pkarr-node-discovery/src/lib.rs:273]: ok

File: iroh-pkarr-node-discovery/src/lib.rs (hunk 1 above)
Comment (rklaehn) [reply iroh-pkarr-node-discovery/src/lib.rs:181]: yeah, you're right.

**Pull request merged**